File: client/src/api/entrenadorService.js
/**
 * Servicio para gestión de Entrenadores
 * Conecta con el endpoint /api/v1/entrenadores del backend
 */

import apiClient from './apiClient'
import { ENDPOINTS } from '../config/api.config'

const EntrenadorService = {
  /**
   * Obtener todos los entrenadores
   * @param {Object} params - Parámetros de filtro y paginación
   */
  getAll: async (params = {}) => {
    const response = await apiClient.get(ENDPOINTS.ENTRENADORES, { params })
    return response.data
  },

  /**
   * Obtener un entrenador por ID
   * @param {number} id - ID del entrenador
   */
  getById: async (id) => {
    const response = await apiClient.get(`${ENDPOINTS.ENTRENADORES}${id}/`)
    return response.data
  },

  /**
   * Crear un nuevo entrenador
   * @param {Object} data - Datos del entrenador
   */
  create: async (data) => {
    const response = await apiClient.post(ENDPOINTS.ENTRENADORES, data)
    return response.data
  },

  /**
   * Actualizar un entrenador existente
   * @param {number} id - ID del entrenador
   * @param {Object} data - Datos a actualizar
   */
  update: async (id, data) => {
    const response = await apiClient.put(`${ENDPOINTS.ENTRENADORES}${id}/`, data)
    return response.data
  },

  /**
   * Eliminar un entrenador
   * @param {number} id - ID del entrenador
   */
  delete: async (id) => {
    const response = await apiClient.delete(`${ENDPOINTS.ENTRENADORES}${id}/`)
    return response.data
  },

  /**
   * Obtener grupos asignados a un entrenador
   * @param {number} id - ID del entrenador
   */
  getGrupos: async (id) => {
    const response = await apiClient.get(`${ENDPOINTS.ENTRENADORES}${id}/grupos/`)
    return response.data
  },

  /**
   * Asignar grupo a entrenador
   * @param {number} entrenadorId - ID del entrenador
   * @param {number} grupoId - ID del grupo
   */
  asignarGrupo: async (entrenadorId, grupoId) => {
    const response = await apiClient.post(`${ENDPOINTS.ENTRENADORES}${entrenadorId}/asignar-grupo/${grupoId}/`)
    return response.data
  },
}

export default EntrenadorService

File: server/basketball/services/entrenador_service.py
"""
Servicio para la gestión de Entrenadores
"""

import logging
from typing import List, Optional, Dict, Any
from django.core.exceptions import ValidationError
from ..dao.entrenador_dao import EntrenadorDAO
from ..models import Entrenador

logger = logging.getLogger(__name__)


class EntrenadorService:
    """Servicio que encapsula la lógica de negocio para Entrenadores"""
    
    def __init__(self):
        self.dao = EntrenadorDAO()
    
    def create_entrenador(self, data: Dict[str, Any]) -> Entrenador:
        """
        Crea un nuevo entrenador.
        
        Args:
            data: Diccionario con los datos del entrenador
            
        Returns:
            Entrenador: Instancia creada
            
        Raises:
            ValidationError: Si los datos no son válidos
        """
        persona_external = data.get('persona_external')
        if not persona_external:
            raise ValidationError("El ID externo de la persona es obligatorio")
            
        if self.dao.exists_persona_external(persona_external):
            raise ValidationError(f"Ya existe un entrenador con el ID externo {persona_external}")
            
        return self.dao.create(**data)
    
    def update_entrenador(self, pk: int, data: Dict[str, Any]) -> Optional[Entrenador]:
        """
        Actualiza un entrenador existente.
        
        Args:
            pk: ID del entrenador
            data: Datos a actualizar
            
        Returns:
            Entrenador | None: Instancia actualizada
        """
        # Validar que no se intente duplicar el persona_external si se está cambiando
        if 'persona_external' in data:
            current = self.dao.get_by_id(pk)
            if current and current.persona_external != data['persona_external']:
                if self.dao.exists_persona_external(data['persona_external']):
                    raise ValidationError(f"Ya existe un entrenador con el ID externo {data['persona_external']}")
        
        return self.dao.update(pk, **data)
    
    def delete_entrenador(self, pk: int) -> bool:
        """
        Elimina (soft delete) un entrenador.
        
        Args:
            pk: ID del entrenador
            
        Returns:
            bool: True si se eliminó correctamente
        """
        return self.dao.soft_delete(pk)
    
    def get_entrenador_by_id(self, pk: int) -> Optional[Entrenador]:
        """Obtiene un entrenador por ID"""
        return self.dao.get_by_id(pk)
    
    def get_all_entrenadores(self) -> List[Entrenador]:
        """Obtiene todos los entrenadores activos"""
        return list(self.dao.get_activos())
    
    def search_entrenadores(self, search_term: str) -> List[Entrenador]:
        """Busca entrenadores por término"""
        return list(self.dao.search_entrenadores(search_term))
